core/src/main/java/org/elasticsearch/action/support/TransportAction.java

```diff
@@ -59,25 +63,46 @@ protected TransportAction(Settings settings, String actionName, ThreadPool threa
     }
 
     public final void execute(Request request, ActionListener<Response> listener) {
+        Task task = taskManager.register("transport", actionName, request);
+
+        ActionListener<Response> listenerWrapper = new ActionListener<Response>() {
```

Comment:
Anyway you can write this so the old listener goes out of scope so its impossible to accidentally use it?